### PR TITLE
PEP proposal for Optimizer class

### PR DESCRIPTION
**Pull request by andyfaff**
@stsievert, @jaimefrio I'm going to write a blurb on why I think Optimizer classes should be introduced. From discussion on scipy-dev this morning it's clear that a case needs to be written, and hopefully you may be able to contribute. This is a location to put that text (not intended for eventual merge into master). Feel free to put requests against this feature branch in `PEP/1-Optimizer.md`. Not wedding to MD, could be rst, etc. If you think it could be done in a better way (e.g. google docs) let me know.

### CODE REVIEW

**Comment by stsievert** [comment]
A useful link: https://github.com/scipy/scipy/pull/6326, which details a similar ODE refactor.

**Comment by stsievert** [comment]
I think we're almost ready to start writing. The outline is looking pretty complete.

I think we should highlight that `minimize` is (effectively) trying to be a class. I fleshed that out more and brought it to the top. Your most recent commit made me realize this is an issue.

I think we should follow the format of Python PEPs, https://github.com/python/peps. We can leave the outline in Markdown.

**Comment by andyfaff** [comment]
>  Some are literal black boxes and implemented in Fortran/C.

I don't think we should say that we're going to extract the Fortran/C code back into Python. That's really well tested over many years, and would be resisted by the community. 

LBFGSB has Fortran, but it returns every step, meaning we can have an Optimizer object. If there are minimizers/solvers that run to completion (leastsq?) within C/Fortran, then I'm not sure that we could translate those easily to Optimizer.

**Comment by andyfaff** [comment]
Yes, `leastsq` is totally Fortran contained - little chance of making an Optimizer out of that, unless it's possible to call Fortran step by step, but that would lead to overhead (possibly unacceptable for many).

BTW I did a little bit of profiling for NelderMead. minimize rosen for starting [1, 2, 3, 4]
fmin is around 24ms
NelderMead is around 29ms.
So the class based approach is around 20% slower. Not sure where that comes from. cython might recover that.

**Comment by andyfaff** [comment]
>* minimize is trying to be a class
>    * method: should be subclasses

Playing devils advocate - the original solvers were `fmin, f_min_bfgs`, etc. Then they were collected together in `minimize` to introduce uniformity, which occurred to a degree. Now we're proposing that they separate again to individual classes. You can see how people can be suspicious of yet another re-write.

**Comment by stsievert** [comment]
> they were collected together in minimize to introduce uniformity, which occurred to a degree

Classes introduce another level of uniformity, correct? Subclasses can inherit from the base class, and with that comes all the parameters and functions. And overridden functions should return values of the same type/shape/etc (though there's no way to really check).

Here are issues that have arisen because of special cases:

* https://github.com/scipy/scipy/issues/8419#issuecomment-365914904: "some optimize.minimize methods modify the parameter vector in-place", fixed by calling `callback` with `x.copy()` in https://github.com/scipy/scipy/issues/8417
* https://github.com/scipy/scipy/issues/8373 "scipy.optimize has broken my trust." which doesn't have the most civil discussion
    * From issue filer: "This has cost me dozens of hours of debugging time, only to learn that **something is wrong with L-BFGS-B that causes the output of options={'disp': True} to not be the the cost function at a given parameter vector.**"
    * From https://github.com/scipy/scipy/issues/8373#issuecomment-363921586 "It'll have to be fixed individually in each minimizer implementation, though, sometimes in Python, sometimes in C, depending on the minimizer."
* https://github.com/scipy/scipy/issues/5832 looks like an issue with Fortran/Python
* https://github.com/scipy/scipy/issues/6019 "minimize_scalar doesn't honor disp option", classes could help with this
* https://github.com/scipy/scipy/issues/7854: "BUG: L-BFGS-B does two more iterations than specified in maxiter"
* https://github.com/scipy/scipy/issues/6673, "return value of scipy.optimize.minimize not consistent for 1D", result inconsistent for different algs
* https://github.com/scipy/scipy/issues/8375, "optimize - check that maxiter is not exceeded."

**Comment by stsievert** [comment]
Can you give the proposal a read over? I added a section on goals and want to make sure we're on the same page.

**Comment by andyfaff** [comment]
One issue is the point 
> hides all details. Some are literal black boxes and implemented in Fortran/C.

I don't think we want to be pulling code out of Fortran and C. As it's written it gives the impression that we're going to do all that calculation in Python. That's an enormous task. The core of the proposal is that pretty much all optimisers have a loop, and that loop can be naturally run as an iterator . So for all the optimizers that have their loop in Python, the loop is extracted and put into `__next__`. For example LBFGSB has a loop, in that loop there is a call to Fortran. That Fortran call will stay as is, but the loop that calls it will now form the core of the `__next__` method.

However, for something like `leastsq` the entire thing is in external code (minpack), calling back into python for func evaluation. At this stage we don't want to be delving into minpack and changing it. However, with the design we have it will allow someone at a future stage to do that.

**Comment by andyfaff** [comment]
I have another idea, which might sway the argument for a proportion of the community.  If we put Optimizer, Function and the Optimizer subclass into Cython it might allow the use of a LowLevelCallable. I know that this provides a big speedup for things like integration where the entire operation then happens in C, it may be that it we could get big speed improvements if everything can happen in Cython, without having to call back into Python.
However, I'm not super familiar with Lowlevelcallable, it'll need some investigating.

**Comment by andyfaff** [comment]
>  bug reports, personal experience and anecdotal evidence.

We need hard evidence here. @jaimefrio, do you reckon you'd be able to chip in here?

**Comment by andyfaff** [comment]
What is meant by a "minimal class interface"?

**Comment by andyfaff** [comment]
Not sure that multiprocessing is immediately introducible as a result of this PR

**Comment by stsievert** [comment]
Thanks for your comments. I've made some edits; could you take another look? 

> Not sure that multiprocessing is immediately introducible as a result of this PR

Maybe we should mention multiprocessing in a "possible future work" section or "possible third party work this enables"? I can point to certain async optimization algorithms that require multiprocessing.

> Optimizer, Function and the Optimizer subclass into Cython 

This would also allow users using Cython to subclass `Optimizer` or `Function`, and get rid of a lot of function call and looping overhead (though I don't know that it's significant).

> bug reports, personal experience and anecdotal evidence. ... We need hard evidence here

We should also ask the managers of other libraries this question. Let's finish a draft of the report then ask them. I've added a note at the top of the draft about what to ask them about.

> It's not obvious what is meant by "black box". Replacing current loops with an iterator is not going to cure that.

What if I don't want my optimizer to do a line search, or what if I want to change some of the hyperparameters in the line search? I think we should have a function `Optimizer.update` (or similar) that updates the current parameter vector `x` (and it should be a line search by default).

> What is meant by a "minimal class interface"?

One that wasn't specific to `minimize` or even SciPy I guess. I think I was imagining some class `Optimizer` with 4 subclasses, one for each of `minimize`, `minimize_scalar`, `root` and `linprog`. I'd imagine this class has a couple basic functions (`update`, `__next__`, ...) that all raise a `NotImplementedError`. 

Then, each of the methods in those functions could subclass from their parents (e.g., `Optimize` is a parent of `Minimize` is a parent of `NealderMead`). This could be used in `solve_ivp` I guess too.

I'm not saying this proposal should implement all 4 subclasses of `Optimizer`, but I think we should lay the groundwork for any future work.

**Comment by stsievert** [comment]
I need to rework the outline. In the second section, we list briefly list the features of this rewrite. The rest of the paper should be expand upon each of those features.

**Comment by andyfaff** [comment]
I've been pondering a bit more. I can understand the reluctance to have 'another API', and whilst that's probably the main issue you're concerned with, I think for me by far the main attraction is that the changes will make maintenance and introducing new features a lot easier. I think we should give the latter more prominence.

One of my other thoughts, and a lesson I learnt with writing DifferentialEvolutionSolver, is that it could be a good idea to keep the Optimizer/Function classes private for a couple of releases. They would still be the core functionality for the current `fmin`, etc, but this approach has the advantage of:

1) being able to tweak implementation of the classes before they become public and it's too late to change anything.
2) being able to tweak implementation if we discover a problem whilst implementing other scalar minimizers.
3) Give us time to convert the other scalar minimizers before a big release.

**Comment by stsievert** [comment]
> keep the Optimizer/Function classes private for a couple of releases

I also think that's a good idea. I hadn't considered it but am on board.

> I think we should give the latter more prominence.

I've moved the "other API" section to the further down.

**Comment by andyfaff** [comment]
I've had a big go at this today. Hopefully it hasn't crossed with anything you've done.
I think a lot of the remaining sections need to be filled out by you, viz the issues you experience with minimize, etc.

**Comment by stsievert** [comment]
I did some work on this earlier this weekend on the [scipep-shorter] branch. I discarded it to think about it more. I'll try to apply those ideas later tonight; I think we could shorten it up and make it more compelling by pointing to specific use cases where this could help (e.g., with the callback flag, stopping early, numerical differentiation)

[scipep-shorter]:https://github.com/andyfaff/scipy/tree/scipep-shorter

**Comment by andyfaff** [comment]
THe document could undoubtedly do with tightening up, as there is some repetition. Could we complete all the sections first though?

1) +Step size selection

THe first sentence doesn't make much sense. I also don't understand what the desired implementation is.  Do you want a line-search method for each Optimizer (possibly a user supplied callable), or the ability to turn the line search on or off, or more control about how line search parameters are specified? At the moment I find it difficult to support this point because I don't understand the vision.

Line search is not used for DifferentialEvolution and Basinhopping.
In LBFGS the line-search is opaque because it's done in Fortran, so it's not applicable there.
Need to be careful with step size as well, it can get confused with finite difference step size.

2) Ease of use - standard interface.
3) ``minimize`` arguments
4) Inheritance for standard interface
5) ``minimize`` is a black box. This is important to provide clarity on. 

""``minimize`` hides a lot of detail"" --> more detail can be discovered by looking at fmin_bfgs docstring, etc. 
"There is no interface to change any of the arguments to these functions or how they operate" --> the options dict can be used to supply solver specific arguments, as would be
accepted by fmin_bfgs.
"We have seen this an issue with" --> give a very specific example. At the moment one example is "expensive functions time-wise", but this doesn't explain what the problem is. The reader has to guess.

6) Expensive functions time-wise
"If function evaluation is expensive time-wise, there may be some more optimizations required based on low level function calls." - it's not clear what this means.
Are you saying that minimization can be tweaked from problem to problem, or are you saying that the algorithms can be improved, or ...?
"currently, this requires rewriting the function and all the functions that call the function desired to be changed." - do you mean the minimizer function, or the objective function.
It's not clear how sklearn improved. They made some changes, but what changes did they make to Newton-CG.
It's not clear that a class-based framework improves the situation here, can't those changes be made directly to the newton-cg minimizer?


7) ``minimize``: class features
I think we should remove this section, because `minimize` isn't going away. Even in the class based Optimizer the OptimizeResult is going to be populated by method specific attributes - e.g. NelderMead doesn't calculate grad, so result.jac will never be populated.
I think what you're trying to get at is that you want to do is supply richer arguments to the underlying minimizer (such as line-search stuff), but `minimize	 (and the underlying solver methods) don't offer that.

8) `minimize` already presents a unifed interface.
I think `minimize` does present a unified interface, and am not sure how it doesn't.
Possibly you're looking to supply more complex arguments to the underlying method, which minimize currently doesn't offer. Inevitably not all those arguments can be usable by different methods. I don't think we can claim that the Optimizers will unify an interface, unless we mean that it unifies the interface for the second level API (fmin, fmin_bfgs, etc). More complex keywords can be added to the Optimizer constructors, but the naming and usage of those needs to be standardised.

**Comment by stsievert** [comment]
> THe document could undoubtedly do with tightening up, as there is some repetition. Could we complete all the sections first though?

I'm going to rewrite the outline, and make sure we're hitting all the points. On first glance, the edits look pretty good though.

**Comment by andyfaff** [comment]
I think we're at the stage now where I'd just like to get it finished. I
can't wait to get it included in scipy.

**Comment by andyfaff** [comment]
To be completed:

1) Third party integration.

2) Please expand on step size selection. Do you want to supply your own step size callable, step size method, or ...? Give me an idea of the implementation.

3) Please explain more clearly in "Expensive functions time-wise" what you would like to do, and how the class framework helps.

4) * gradient or Hessian approximation
This point isn't explained. Please explain what you'd like to do and how the class framework helps.

5)  ``minimize`` is similar to root finding and linear programs
needs completing.

File: PEP/1-Optimizer.rst [hunk 1]
```diff
@@ -654,13 +654,10 @@ behaviour is achieved.
...
-Initial benchmarking of ``Optimizer`` for various problems are listed below.
-For extremely fast objective functions there is a ~20% time slowdown for class
-based operation compared to existing code. However, once the objective function
-takes a significant time to evaluate (e.g. > 40us per call) then the introduced
-overhead is negligible.
-
-This small slowdown for some problems must be considered against a wider context:
+Initial benchmarking of ``Optimizer`` for the ``optimize`` benchmark suite are
+listed below. There is a slowdown for those methods (``nelder-mead`, ``bfgs``,
+``l-bfgs-b``) using an ``Optimizer``. However, this slowdown must be considered
```
Comment (stsievert) [inline PEP/1-Optimizer.rst]: Can we quantify the slowdown? It sounds really slow.

**Comment by andyfaff** [comment]
I've just put in a table of the asv benchmarking. I'd be interested to hear your reaction on the relative timings. I'm not sure there's much that can be done to improve speed, but I still think the 'premature optimization' argument applies.

File: PEP/1-Optimizer.rst [hunk 2]
```diff
@@ -654,19 +659,150 @@ behaviour is achieved.
...
+There is some degree of uncertainty in the actual timings. For example,
+'rosenbrock_tight + CG', or 'LJ + COBYLA' have a slowdown factor of ~1.23, yet
+neither of those solvers use an ``Optimizer``. This may indicate that more
+repeats are necessary to obtain more accurate timings.
+
++---+------------+------------+-------+------------------------+----------------+-----------+
+|   | before     | after      | ratio |  problem               | method         | resource  |
+|   | [3cb29828] | [1bbf5b1a] |       |                        |                |           |
++===+============+============+=======+========================+================+===========+
+| + | 544.76μs   | 1.02ms     | 1.87  | sin1d                  | nelder-mead    | mean_time |
```
Comment (stsievert) [inline PEP/1-Optimizer.rst]: Instead of the mean time, we should use the minimum time. Other operating systems get in the way and can slow down the function, and the function will execute no faster than the minimum time. When comparing performance that's more reasonable.

I would be interested in seeing a graph of this.

**Comment by stsievert** [comment]
Thanks for generating these data. This is really useful.

I'd also be interested in seeing some profiling of the new implementation, maybe a chart of `function_time` vs `slowdown_ratio` for different problems and optimizers?

If we see significant slowdown (more than 20% for some function?) it might be worth speed optimization. This is where the profiling of `function_time` vs `slowdown ratio` would be useful.

Why is the new method faster in (many) cases? Have you verified that `nfev`/`njev`/the number of optimization steps are the same?

File: PEP/1-Optimizer.rst (hunk 2 above)
Comment (stsievert) [reply PEP/1-Optimizer.rst]: Also, could you put the timing script in the repo? I'd be curious to see, and I think others would care too.

**Comment by andyfaff** [comment]
I think NelderMead and LBFGSB use the same nfev. I have to check BFGS that there aren't changes in nfev.  A more rigorous check is required there (esp BFG), to ensure the solution trajectories are the same.
Getting the rst table into shape was quite an effort, I wouldn't want to be doing that in a hurry.

I've had a brief go at line-profiling, but I'm not sure that we're going to get much joy (willing to be proved wrong).

The timing script is `python runtests.py --bench-compare master optimize` when on the OPTIMIZE branch. There are a couple of adjustments I had to make to the benchmark suite to get to run nelder-mead.

We have to remember that we're talking us and ms here. For my problems the average function evaluation is between 500us to 7ms for the simpler systems. That's going to dwarf class overhead. For people trying to do high speed trading it might be an issue.

**Comment by andyfaff** [comment]
Take a look at this [gist](https://gist.github.com/andyfaff/9a013ee2f0fe30afa978335a9c975ca2). It shows the line profiling for NelderMead. The total time for the __next__ method from __call__ is ~0.05s. However, the total time spent in __next__ is apparently ~0.04s. This means there must be 0.01s overhead in finding and calling __next__ from __call__ (unless line_profiler can't be used to do that calculation). Thats approximately 20% overhead.

**Comment by andyfaff** [comment]
Scott, what do estimate the timeline is to complete the remaining sections (as outlined in my `To be completed:` comment)? I think those sections have to be up to you, as I'm not clear on what is required in there.
Do you think we'd be able to provide the PEP to the scipy community next week?

**Comment by stsievert** [comment]
I've made some edits. I think we should ask other library maintainers for their opinions – if positive, that will help convince others.

I think it's good enough, though nowhere near perfect.

**Comment by andyfaff** [comment]
If we waited for perfect we'd be a long time.

**Comment by andyfaff** [comment]
I think it'll be better served going straight to scipy-dev, they can always suggest we go out to the projects afterwards.

**Comment by stsievert** [comment]
Numba might be worth mentioning. There's discussion on adding it as a dependency ([scipy-dev post], [a good overview], [numba 2788], [pydata/sparse 126]), and I'd imagine we could use this to speed up access to class attributes/for loops.

[numba 2788]:https://github.com/numba/numba/issues/2788
[scipy-dev post]:https://mail.python.org/pipermail/scipy-dev/2018-March/022576.html
[pydata/sparse 126]:https://github.com/pydata/sparse/issues/126
[a good overview]:http://matthewrocklin.com/blog/work/2018/01/30/the-case-for-numba

**Comment by andyfaff** [comment]
As with anything I'd like to see the evidence for that before concurring. For example, I thought Cython would speed up this PR, but there were no significant gain when I experimented with that. I hope that the real gain would be to use LowLevelCallable from Cython to reduce overhead from Python function calls.

**Comment by andyfaff** [comment]
@rgommers, I made the formatting changes you suggested, such as the header table. I think we're happy with the PEP to be placed in a scipy/* repo at this point.

**Comment by stsievert** [comment]
@rgommers I'm also satisfied with the proposal (though it's not perfect and could use improvement).

I think SEP or SPEP are the best acronyms. Maybe SciPEP?

**Comment by rgommers** [comment]
Thanks, looks good. I'll have a look at creating that repo now.

**Comment by rgommers** [comment]
Okay either https://github.com/numpy/neps is still in flux, or we are keeping the sources for NEPs within the numpy repo itself after all, I can't remember. If the answer is the latter, let's put SciPEPs at `doc/source/scipeps` in the SciPy repo.
